src/execution/execute-afm.js

```diff
@@ -0,0 +1,87 @@
+// Copyright (C) 2007-2014, GoodData(R) Corporation. All rights reserved.
+import { get, mapValues, clone } from 'lodash';
+import invariant from 'invariant';
+import { ajax, post, parseJSON } from '../xhr';
+import { queryString } from '../util';
+
+const PAGE_SIZE = 500;
+const DEFAULT_DIMENSION_COUNT = 2;
+
+function getDimensionality(execution) {
+    const dimensions = get(execution, 'execution.resultSpec.dimensions');
+    return dimensions
+        ? dimensions.length
+        : DEFAULT_DIMENSION_COUNT;
+}
+
+function getLimit(offset) {
+    return Array(offset.length).fill(PAGE_SIZE);
+}
+
+function fetchExecutionResult(pollingUri, offset) {
+    const query = { limit: getLimit(offset), offset };
+    const uri = pollingUri + queryString(mapValues(query, arr => arr.join(',')));
+    return ajax(uri, { method: 'GET' }).then((r) => {
+        if (r.status === 204) {
+            const err = new Error('Loading executeAfm failed: 204 No Content');
+            err.response = r;
+            throw err;
+        }
+        return r.json();
+    });
+}
+
+// works only for one or two dimensions
+export function mergePageData(resultSoFar, { paging, data }) {
+    const rowOffset = paging.offset[0];
+    if (resultSoFar.data[rowOffset]) { // appending columns to existing rows
+        for (let i = 0; i < data.length; i += 1) {
+            resultSoFar.data[i + rowOffset].push(...data[i]);
+        }
+    } else { // appending new rows
+        resultSoFar.data.push(...data);
+    }
+    return resultSoFar;
+}
+
+export function nextPageOffset({ offset, total }) {
+    const newOffset = clone(offset);
+    const maxDimension = offset.length - 1;
+    // we need last dimension first (aka columns, then rows) to allow array appending in merge fnc
+    for (let i = maxDimension; i >= 0; i -= 1) {
+        if (newOffset[i] + PAGE_SIZE < total[i]) {
+            newOffset[i] += PAGE_SIZE;
+            return newOffset;
+        }
+        newOffset[i] = 0;
+    }
+    return false;
+}
+
+function getOnePage(pollingUri, offset, resultSoFar = false) {
```

Comment:
Would it make sense to rename the `resultSoFar` – which seemed like boolean flag to me at first to me – to `prevResult = null`?

Reply:
Much better 👍